### PR TITLE
Parser

### PR DESCRIPTION
**Pull request by shunyao643**
Add Parser Implementation for update commands.

@yuhaochua, this is a draft of how it will look like (did not manage to test edge cases yet), feel free to work off this for now for UpdateCommand.

### CODE REVIEW

File: src/main/java/seedu/parser/Parser.java [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+                args = prepareDelete(commandAndArgument.get(1));
+                return new DeleteCommand(args);
+            } catch (IncompleteCommandException e) {
+                return new IncorrectCommand(DeleteCommand.COMMAND_WORD + DeleteCommand.COMMAND_DESCRIPTION);
+            }
+            break;
+        case UpdateCommand.COMMAND_WORD:
+            try {
+                args = extractArguments(commandAndArgument.get(1));
+                UpdateCommand updateCommand = new UpdateCommand(serialNumber);
```
Comment (yuhaochua) [inline src/main/java/seedu/parser/Parser.java:81]: Maybe we can define this portion in a prepareUpdate method instead?

File: src/main/java/seedu/parser/Parser.java [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+                for (String s : args) {
+                    int delimiterPos = s.indexOf('/');
+                    String argType = s.substring(0, delimiterPos);
+                    String argValue = s.substring(0, delimiterPos + 1);
```
Comment (yuhaochua) [inline src/main/java/seedu/parser/Parser.java:85]: Shouldn't the argValue be from delimiterPos + 1 till the end?

**Review by yuhaochua — APPROVED** [review]

File: src/main/java/seedu/parser/Parser.java [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+                    switch (argType) {
+                    case 'n':
+                        updateCommand.setUpdateName(argValue);
+                        break;
+                    case 'pd':
```
Comment (yuhaochua) [inline src/main/java/seedu/parser/Parser.java:90]: We cannot use the single quotes for string values. Maybe we can change all cases to have double quotes instead?

File: src/main/java/seedu/parser/Parser.java [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+            matcher.usePattern(ARGUMENT_TRAILING_FORMAT);
+            matcher.find(lastIndex);
+            splitArguments.add(matcher.group());
+        } catch (IllegalStateException e) {
+            throw new IncompleteCommandException("No parameters found!");
+        }
+        return splitArguments;
+    }
+
+    protected ArrayList<>
```
Comment (yuhaochua) [inline src/main/java/seedu/parser/Parser.java:140]: This line should be deleted?

File: src/main/java/seedu/parser/Parser.java (hunk 1 above)
Comment (yuhaochua) [inline src/main/java/seedu/parser/Parser.java:81]: the serialNumber here is not defined

File: src/main/java/seedu/parser/Parser.java [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+        ArrayList<String> commandAndArgument = null;
+        ArrayList<String> args = null;
+        try {
+            commandAndArgument = splitCommandTerm(userInput);
+        } catch (IncompleteCommandException e) {
+            return new IncorrectCommand(MESSAGE_INCOMPLETE_COMMAND_MISSING_DELIMITER);
+        }
+
+        switch (commandAndArgument.get(0)) {
+        case AddCommand.COMMAND_WORD:
```
Comment (yuhaochua) [inline src/main/java/seedu/parser/Parser.java:54]: There is a missing case for ListCommand.

File: src/main/java/seedu/parser/Parser.java [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+     * 2. <code> type </code>: String representation of enumerated class
+     * 3. <code> cost </code>: String representation of double value, "$" optional but "," delimiter forbidden
+     * 4. <code> purchasedFrom </code>: String of vendor name, suggest adhering to one consistent naming scheme
+     * 5. <code> purchasedDate </code>: String representation for now, possibility for future support
+     *
+     * @param args
+     * @return ArrayList of arguments
+     * @throws IncompleteCommandException if no match found
+     */
+    protected ArrayList<String> prepareAdd(String args) throws IncompleteCommandException {
```
Comment (yuhaochua) [inline src/main/java/seedu/parser/Parser.java:157]: This method does not split the string properly into the individual attributes of the equipment

Comment (shunyao643) [reply src/main/java/seedu/parser/Parser.java:157]: Debugging this now

File: src/main/java/seedu/parser/Parser.java (hunk 4 above)
Comment (shunyao643) [reply src/main/java/seedu/parser/Parser.java:140]: My bad, missed this out!

File: src/main/java/seedu/parser/Parser.java (hunk 3 above)
Comment (shunyao643) [reply src/main/java/seedu/parser/Parser.java:90]: Thanks!

File: src/main/java/seedu/parser/Parser.java (hunk 2 above)
Comment (shunyao643) [reply src/main/java/seedu/parser/Parser.java:85]: Yep, will fix!

File: src/main/java/seedu/parser/Parser.java (hunk 1 above)
Comment (shunyao643) [reply src/main/java/seedu/parser/Parser.java:81]: Should I pass in a serialNumber into the constructor? Since any update will require that anyway.

File: src/main/java/seedu/parser/Parser.java (hunk 5 above)
Comment (shunyao643) [reply src/main/java/seedu/parser/Parser.java:54]: Cheers, will fix

File: src/main/java/seedu/parser/Parser.java (hunk 1 above)
Comment (yuhaochua) [reply src/main/java/seedu/parser/Parser.java:81]: Yes sure that'll work! But just have to define it before passing it into the constructor, thanks!

Comment (shunyao643) [reply src/main/java/seedu/parser/Parser.java:81]: My original thought process for this was to go with keeping the arguments as is because we don't know what order the arguments could be in, so I will have to iterate through all of them anyway. 

If we're looking to implement it with serialNumber in constructor I think I'll need a prepareUpdate method to find it, otherwise I'm thinking at the end of the for loop, I throw an exception if there isn't such a serialNumber found. Do you have a preference for either implementation? @yuhaochua

Comment (yuhaochua) [reply src/main/java/seedu/parser/Parser.java:81]: I'm fine with either implementation! My bad, actually I was referring to shifting the whole switch case statements below into a prepareUpdate method so it doesn't get too long here.

Comment (shunyao643) [reply src/main/java/seedu/parser/Parser.java:81]: My bad haha, missed out quite a lot stuff trying to push this one out

Comment (shunyao643) [reply src/main/java/seedu/parser/Parser.java:81]: Okay, I've thought about the constructor part. I think whichever class is handling the duplicate entries (e.g. two purchase dates) should also handle missing serial number.

Sure thing, I can refactor prepareUpdate
